### PR TITLE
Made VerseControl usable by Paratext plugins

### PR DESCRIPTION
**Pull request by tombogle**
Added IScrVerseRef interface (largely extracted from VerseRef)
Made VerseRef class implement new IScrVerseRef interface
Changed VerseControl to use IScrVerseRef and not depend directly on ScrVers

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1086)

### CODE REVIEW

File: SIL.Scripture/IScrVerseRef.cs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+		/// <summary>
+		/// Gets whether the versification of this reference has verse segments
+		/// </summary>
+		bool VersificationHasVerseSegments { get; }
+
+		/// <summary>
+		/// Returns verse ref with no bridging, but maintaining segments like "1a".
+		/// </summary>
+		public IScrVerseRef UnBridge();
```
Comment (ermshiperete) [inline SIL.Scripture/IScrVerseRef.cs:47]:
```suggestion
		IScrVerseRef UnBridge();
```

File: SIL.Scripture/IScrVerseRef.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+		/// <summary>
+		/// Tries to move to the previous book in the entire canon superset.
+		/// </summary>
+		/// <returns>true if successful</returns>
+		bool PreviousBook();
+
+		/// <summary>
+		/// Tries to move to the next chapter.
+		/// </summary>
```
Comment (ermshiperete) [inline SIL.Scripture/IScrVerseRef.cs:94]:
```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```

File: SIL.Scripture/IScrVerseRef.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		/// <returns>true if successful</returns>
+		bool NextChapter();
+
+		/// <summary>
+		/// Tries to move to the previous chapter.
+		/// </summary>
```
Comment (ermshiperete) [inline SIL.Scripture/IScrVerseRef.cs:106]:
```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```

File: SIL.Scripture/IScrVerseRef.cs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		/// <returns>true if successful</returns>
+		bool PreviousChapter();
+
+		/// <summary>
+		/// Tries to move to the next verse (or verse segment, if available in the current versification).
+		/// </summary>
```
Comment (ermshiperete) [inline SIL.Scripture/IScrVerseRef.cs:118]:
```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```

File: SIL.Scripture/IScrVerseRef.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+		/// <returns>true if successful, false if at end of scripture</returns>
+		bool NextVerse();
+
+		/// <summary>
+		/// Tries to move to the previous verse (or verse segment, if available in the current versification).
+		/// </summary>
```
Comment (ermshiperete) [inline SIL.Scripture/IScrVerseRef.cs:130]:
```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```

File: SIL.Scripture/VerseRef.cs [hunk 6]
```diff
@@ -378,6 +377,13 @@ public string[] GetSegments(string[] defaultSegments)
 			return segsForThisVerse ?? defaultSegments;
 		}
 
+		public void AdvanceToLastSegment()
+		{
+			string[] segments = GetSegments(null);
+			if (segments != null && segments.Length > 0)
```
Comment (ermshiperete) [inline SIL.Scripture/VerseRef.cs:383]: or

```suggestion
			if (segments?.Length > 0)
```

File: SIL.Scripture/VerseRef.cs [hunk 7]
```diff
@@ -808,6 +813,11 @@ public bool NextChapter(BookSet present, bool skipExcluded = false)
 			return true;
 		}
 
+		public bool NextChapter(BookSet present)
+		{
+			return NextChapter(present);
```
Comment (ermshiperete) [inline SIL.Scripture/VerseRef.cs:818]: Won't that cause an infinite loop? I guess this should probably be:

```suggestion
			return NextChapter(present, false);
```

**Review by ermshiperete — APPROVED** [review]

**Review by tombogle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1086)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @ermshiperete and @FoolRunning)

___
*[SIL.Scripture/IScrVerseRef.cs, line 47 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-Mfi_3rdgwwOkbBtvTI5-r1-47:-MfnQWV6DM9wK2-ucfar:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e9040b9aebae2a71aae0b37159ac0a6e1fe9ec15/SIL.Scripture/IScrVerseRef.cs#L47)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		IScrVerseRef UnBridge();
```
</blockquote></details>

Done.
___
*[SIL.Scripture/IScrVerseRef.cs, line 94 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-Mfi_3rdgwwOkbBtvTI5-r1-94:-MfnQYjL6EV2kJbdN6wt:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e9040b9aebae2a71aae0b37159ac0a6e1fe9ec15/SIL.Scripture/IScrVerseRef.cs#L94)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```
</blockquote></details>

Done.
___
*[SIL.Scripture/IScrVerseRef.cs, line 106 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-Mfi_3rdgwwOkbBtvTI5-r1-106:-MfnQZCCBs-DgMEpT3V2:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e9040b9aebae2a71aae0b37159ac0a6e1fe9ec15/SIL.Scripture/IScrVerseRef.cs#L106)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```
</blockquote></details>

Done.
___
*[SIL.Scripture/IScrVerseRef.cs, line 118 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-Mfi_3rdgwwOkbBtvTI5-r1-118:-MfnQZaU56g3CZ_pEkFN:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e9040b9aebae2a71aae0b37159ac0a6e1fe9ec15/SIL.Scripture/IScrVerseRef.cs#L118)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```
</blockquote></details>

Done.
___
*[SIL.Scripture/IScrVerseRef.cs, line 130 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-Mfi_3rdgwwOkbBtvTI5-r1-130:-MfnQ_9947H06UiQT_OJ:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e9040b9aebae2a71aae0b37159ac0a6e1fe9ec15/SIL.Scripture/IScrVerseRef.cs#L130)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// </summary>
		/// <param name="present">Set of books present or selected.</param>
```
</blockquote></details>

Done.
___
*[SIL.Scripture/VerseRef.cs, line 383 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-Mfi_3rdgwwOkbBtvTI3-r1-383:-MfnQ_wU-SDrb1KDWGcw:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e9040b9aebae2a71aae0b37159ac0a6e1fe9ec15/SIL.Scripture/VerseRef.cs#L383)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

or

```suggestion
			if (segments?.Length > 0)
```
</blockquote></details>

Done.
___
*[SIL.Scripture/VerseRef.cs, line 818 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-Mfi_3rdgwwOkbBtvTI3-r1-818:-MfnQamqDY0oNB6OmgDe:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e9040b9aebae2a71aae0b37159ac0a6e1fe9ec15/SIL.Scripture/VerseRef.cs#L818)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Won't that cause an infinite loop? I guess this should probably be:

```suggestion
			return NextChapter(present, false);
```
</blockquote></details>

Done.

**Review by FoolRunning — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1086)* status: all files reviewed, 8 unresolved discussions (waiting on @ermshiperete and @tombogle)

___
*[SIL.Scripture/VerseRef.cs, line 17 at r2](https://reviewable.io/reviews/sillsdev/libpalaso/1086#-MfrPYdVBnyfmLn2-qum:-MfrPYdW7CAyCAq9-yIj:b9zwgt8) ([raw file](https://github.com/sillsdev/libpalaso/blob/218eb2edac900d87243f56df5472cf872f46254d/SIL.Scripture/VerseRef.cs#L17)):*
> ```Smalltalk
> 	{
> 		#region Constants
> 		[PublicAPI]
> ```

For future reference, you can just add PublicAPI to the class and all the public members will be considered PublicAPI as well.

**Comment by tombogle** [comment]
Thanks, good to know.

**Pull request merged**